src/main/java/com/google/devtools/build/lib/skyframe/FilesystemValueChecker.java

```diff
@@ -417,6 +422,14 @@ private boolean actionValueIsDirtyWithDirectSystemCalls(ActionExecutionValue act
         try {
           ArtifactFileMetadata fileMetadata =
               ActionMetadataHandler.fileMetadataFromArtifact(file, null, tsgm);
+          FileArtifactValue fileValue = actionValue.getArtifactValue(file);
+          boolean lastSeenRemotely = fileValue != null && fileValue.isRemote();
+          if (!fileMetadata.exists() && lastSeenRemotely) {
```

Comment:
What are the expected semantics when:

1. A file is created remotely
2. The file is later materialized
3. The file is then deleted by `rm`

Is Bazel expected to notice the deletion in that case?

Reply:
> Is Bazel expected to notice the deletion in that case?

It wouldn't expect deletion. Do you think that having this behavior is important?